core/src/main/java/kafka/server/handlers/DescribeTopicPartitionsRequestHandler.java

```diff
@@ -0,0 +1,137 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.server.handlers;
+
+import kafka.network.RequestChannel;
+import kafka.server.AuthHelper;
+import kafka.server.KafkaConfig;
+import kafka.server.MetadataCache;
+import kafka.server.metadata.KRaftMetadataCache;
+import kafka.server.metadata.ZkMetadataCache;
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.errors.InvalidRequestException;
+import org.apache.kafka.common.message.DescribeTopicPartitionsRequestData;
+import org.apache.kafka.common.message.DescribeTopicPartitionsResponseData;
+import org.apache.kafka.common.message.DescribeTopicPartitionsResponseData.DescribeTopicPartitionsResponsePartition;
+import org.apache.kafka.common.message.DescribeTopicPartitionsResponseData.DescribeTopicPartitionsResponseTopic;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.requests.DescribeTopicPartitionsRequest;
+import org.apache.kafka.common.resource.Resource;
+import scala.collection.JavaConverters;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+import static org.apache.kafka.common.acl.AclOperation.DESCRIBE;
+import static org.apache.kafka.common.resource.ResourceType.TOPIC;
+
+public class DescribeTopicPartitionsRequestHandler {
+    MetadataCache metadataCache;
+    AuthHelper authHelper;
+    KafkaConfig config;
+
+    public DescribeTopicPartitionsRequestHandler(
+        MetadataCache metadataCache,
+        AuthHelper authHelper,
+        KafkaConfig config
+    ) {
+        this.metadataCache = metadataCache;
+        this.authHelper = authHelper;
+        this.config = config;
+    }
+
+    public DescribeTopicPartitionsResponseData handleDescribeTopicPartitionsRequest(RequestChannel.Request abstractRequest) {
+        if (metadataCache instanceof ZkMetadataCache) {
+            throw new InvalidRequestException("ZK cluster does not handle DescribeTopicPartitions request");
+        }
+        KRaftMetadataCache kRaftMetadataCache = (KRaftMetadataCache) metadataCache;
+
+        DescribeTopicPartitionsRequestData request = ((DescribeTopicPartitionsRequest) abstractRequest.loggableRequest()).data();
+        Set<String> topics = new HashSet<>();
+        boolean fetchAllTopics = request.topics().isEmpty();
+        DescribeTopicPartitionsRequestData.Cursor cursor = request.cursor();
+        if (fetchAllTopics) {
+            if (cursor != null) {
+                // Includes the cursor topic in case the cursor topic does not exist anymore.
+                topics.add(cursor.topicName());
```

Comment:
Instead of doing this, can we get all topics that are >= cursor and then check (after sorting) if the cursor topic is present and reset partition index to 0 if it's not?  This would also handle the unauthorized case.

Reply:
Good catch, if the cursor topic is unauthorized, the first partition id is messed up.